configfile: "config/config.yaml"

include: "rules/files_prep.smk"
include: "rules/alignment.smk"
include: "rules/QC.smk"
include: "rules/analysis.smk"
include: "rules/variant_call_phasing.smk"
include: "rules/ASE.smk"
	
SMP = [0]

myoutput=list()

if "gex_bulk" in config["tech"]:
    myoutput.append(expand("results/{tec}/ASE{chrom}_bulk",chrom=config["chromosomes_to_phase"],tec="gex_bulk"))
if "atac" in config["tech"]:
    myoutput.append(expand("results/{tec}/ASE{chrom}",chrom=config["chromosomes_to_phase"],tec="atac"))
if "gex" in config["tech"]:
    myoutput.append(expand("results/{tec}/ASE{chrom}",chrom=config["chromosomes_to_phase"],tec="gex"))


#functions 

def get_file_names(wildcards):
    ck_output = checkpoints.split_bam.get(**wildcards).output[0]
    global SMP
    SMP, = glob_wildcards(os.path.join(ck_output, "cluster_{sample}.bam"))
    return expand("results/{tech}/data_by_clusters/cluster_{SAMPLE}.bam", SAMPLE=SMP,tech=wildcards.tec)

#rules and checkpoints 

rule all:
    input:
        myoutput,
        expand("results/phased/chr{chrom}_phased.vcf.gz",chrom=config["chromosomes_to_phase"])
